internal/core/processor/evolution/processor.go

```diff
@@ -98,6 +102,21 @@ func (p *processor) ProcessEvoBlockRange(ctx context.Context, startingBlock, las
 	tx := p.stateService.NewTransaction()
 	defer tx.Discard()
 
+	for {
+		var ok bool
+		ok, err := p.hasBlockFinalize(big.NewInt(int64(lastBlock)))
+		if err != nil {
+			slog.Error("error occurred while checking latest finalized block", "err", err.Error())
+			return err
+		}
+		if ok {
+			break
+		}
+		slog.Debug("block not finalized, waiting for finality", "block", lastBlock)
+
+		time.Sleep(time.Millisecond * 500)
```

Comment:
this sleep and infinite loop potentialy won't allow you to turn off the node gracefully if you don't get last finalized block. you should use.

```go
func WaitBeforeNextScan(ctx context.Context, waitingTime time.Duration) {
	timer := time.NewTimer(waitingTime)
	select {
	case <-ctx.Done():
		timer.Stop()
	case <-timer.C:
	}
}
```


it is defined in worker.

Reply:
I'll be substituting it; however, even though this approach is an improvement, the existing method shouldn't lead to an infinite loop. We're requesting a block number, not a block hash, so eventually, a block with the specified number X will be generated.